hollow/src/main/java/com/netflix/hollow/tools/util/ObjectInternPool.java

```diff
@@ -0,0 +1,68 @@
+package com.netflix.hollow.tools.util;
+
+import java.util.Map;
+import java.util.HashMap;
+
+// This class memoizes types by returning references to existing objects, or storing
+// Objects if they are not currently in the pool
+public class ObjectInternPool {
+    final private GenericInternPool<Integer> integerInternPool = new GenericInternPool<>();
+    final private GenericInternPool<Float> floatInternPool = new GenericInternPool<>();
+    final private GenericInternPool<Double> doubleInternPool = new GenericInternPool<>();
+    final private GenericInternPool<Long> longInternPool = new GenericInternPool<>();
+
+    public Object intern(Object objectToIntern) {
+        if(objectToIntern==null) {
+            throw new IllegalArgumentException("Cannot intern null objects");
+        }
+
+        // Automatically handles booleans and integers within cached range
+        if(objectAutomaticallyCached(objectToIntern)) {
+            return objectToIntern;
+        }
+
+        if(objectToIntern instanceof Float) {
+            return floatInternPool.intern((Float) objectToIntern);
+        } else if(objectToIntern instanceof Double) {
+            return doubleInternPool.intern((Double) objectToIntern);
+        } else if(objectToIntern instanceof Integer) {
+            return integerInternPool.intern((Integer) objectToIntern);
+        } else if(objectToIntern instanceof Long) {
+            return longInternPool.intern((Long) objectToIntern);
+        } else if(objectToIntern instanceof String) {
+            // Use Java's builtin intern function
+            return ((String) objectToIntern).intern();
+        } else {
+            String className = objectToIntern.getClass().getName();
+            throw new IllegalArgumentException("Cannot intern object of type " + className);
+        }
+    }
+
+    private boolean objectAutomaticallyCached(Object objectToIntern) {
+        if(objectToIntern instanceof Boolean) {
+            return true;
+        } else if(objectToIntern instanceof Integer) {
+            return -128 <= (Integer) objectToIntern && (Integer) objectToIntern <= 127;
+        }
+        return false;
+    }
+}
+
+class GenericInternPool<T> {
+    private final Map<T, T> pool = new HashMap<>();
+
+    public T intern(T object) {
+        if (object == null) {
+            throw new IllegalArgumentException("Cannot intern null objects");
+        }
+
+        synchronized (pool) {
```

Comment:
> I tried to sample the results with one of our ~1G datasets.. I didn't get a performance improvement in that, I suspect it may be due to this synchronization. This could be optimized e.g. do an optimistic get, lock stripping. Or just use ConcurrentHashMap? I dont think there would be much more memory overhead.

I'm not sure if synchronization is the cause of the lower performance as each type index has its own pool, and I believe there is no more than one thread operating on a type index at a time. With that said, I didn't run a profiler after the addition of the intern pool, so it may still be the cause.

Can you elaborate on how you measured the performance improvement, as well as if the dataset is publicly available? If I can replicate, I can figure out the issue with a profiler.

> There was a little bug that came up in the primary key field parsing. The primary key referenced one field, like "id.value", this patch should fix it-

Good catch! I'll add that to the PR. 


> The memory consumption optimization from interning everything has an interesting aspect: for datasets that don't have much repetition in Integer/Long/Float/Double values the overhead of interning adds up.. I think 16-32 bytes per object that isn't repeated.

Can you explain why this is the case? My idea was that Java will store only one instance of an object, as the hashmap in `HollowOrdinalMapper` will contain a reference to the object stored in the memoized pool. 

All references for equal objects in the pool's hashmap and ordinalmapper's hashmap point to the same object. We _are_ storing an extra reference to the object, but this is also the case with Java's string intern method. If I'm misunderstanding anything (your thoughts or the Java language), please let me know.

Thanks for taking a look!

Reply:
Yes re: the 16-32 bytes overhead, I was thinking of the references in HashMap. HashMap.Node looks like 
```
        final int hash;
        final K key;
        V value;
        Node<K,V> next;
```
This cost might be bearable and like you said unavoidable if we memoize objects.. We may need to add reference counting and drop objects from the intern pool when old history states are removed to avoid memory creep for long running history servers. 

>If the previous Hollow memoization implementation avoids this per-object overhead, please let me know.

I guess it's not really interning/memoizing(sorry for my poor use of that term), it's deduping field values within the specific historic state by adding records to a write state engine- the  memory overhead for storing those values is expected to be lower because of use of ordinals and variable length encoding into byte arrays, but we dont dedupe values across versions so the overall memory utilization may be higher if there is repetition in values across states. So theres that tradeoff there.. I wasn't aware you already chatted with Drew- if he suggested memoization lets go ahead and do that- I can also compare the memory footprint when comparing performance to validate with real world data.

Reply:
+1 the primary benefit of faster snapshot/delta processing time outweighs the desire to optimize memory util. In production we dont care about squeezing as many historic states as possible, rather we need to be able to look back X no. of states per dataset where X is different for different datasets. So optimal memory util translates to slightly less RAM costs, but I dont think the difference between the 2 approaches is that drastic, esp. considering it depends on nature of changes in the underlying data.

That said, my hunch is that an ever-growing interning pool could pose a problem for running these history servers in a stable manner- but let me see with data what that creep actually looks like, and if that extra bookkeeping could be valuable.

Reply:
[This tweak](https://github.com/Netflix/hollow/commit/488f9c5470c12753e153e4e474a7ce2c637902c5) for handling the edge case where some fields appeared as "null"- to preserve behavior where `addTypeIndex` could be called called additively

Reply:
Thanks for the data, I have a few thoughts on this. I believe Drew mentioned that the current implementation of `HollowHistory` uses a series of deltas, rather than snapshots. If this is the case, it'll be hard to prevent unrestricted heap growth because all of the field objects for ordinals in the state need to be accessible.

I'm not particularly concerned with the fact that the heap is growing over time, this should be expected as we need to be able to retrieve the values of all historical states. I may be mistaken about this, though, especially if snapshots are used (note: I'm referring to snapshots within HollowHistory, not when processing the actual dataset). Also, Drew recommended the use of the `intern()` function, which keeps strings in the heap for the lifetime of the program.

The current intern pool has a pretty simple optimization I can implement to reclaim memory on snapshots. If snapshots are used, this could reclaim large amounts of memory.

Can you please test the heap growth without memoization by changing this to `indexFieldObjectMapping.get(index)[i] = objectToStore;`?
https://github.com/Netflix/hollow/blob/a564b70711e057e95ce7cc7e8c99dffb711f8640/hollow/src/main/java/com/netflix/hollow/tools/history/keyindex/HollowOrdinalMapper.java#L126

Let me know if you think it'd be worth having a quick call to discuss.

Reply:
It's true our implementation relies on listening to delta updates. But also see `HollowHistory::removeHistoricalStates`. Basically, once sufficient no. of historic states have been indexed you can enter a "sliding window" mode where oldest states are evicted in order to accommodate new versions thereby keeping the heap util somewhat stable. It is desirable to have x versions / days of history available at any given time, as opposed of having to wait for a newly initialized history server to catch up (which could be hours for large datasets with high entropy).

Reply:
I see, would it be worth having a 20-30min call to discuss these changes? I've received different suggestions from you and Drew (who said that ordinals in the mapping should remain forever), so I think it may be helpful for us to sync and get on the same page.

Reply:
Yes, it would be good to sync up. Would Wed or Thu after 2 PM PST work? Is the email listed on your github profile a good one to send the meeting details out to?

Reply:
Yes, that works well.

Reply:
Resolving due to discussing on call and deciding that further tests will be done. We can followup later on if more discussion is needed.